### PR TITLE
[BUG FIX] Fixes a bug where teachers couldn't view any programs of non-class members

### LINKED ISSUE #2174: [BUG] Check fix #2173
When you are back, you will want to see if my hotfix is sensible :)

### LINKED ISSUE #2246: [BUG] Public programs from users within a class can't be edited
**Describe the bug**
This is a pretty unique bug. We should allow the following two situations:

- Teachers can view a student profile -> with **all** their programs (also non-shared ones)
- Other user's can view all shared programs
   - They are also allowed to edit them for themselves

However, currently we perform a faulty check. Programs by users who are currently in a class are not allowed to be visited by any user. Regardless if these programs are public or not. This might have something to do with a faulty check if the visitor is also the teacher of the user. We have to improve this by stating:

_You can visit this program if either of the following is true:_

- You are the admin
- You are the teacher of the owner of the program
- The program is public

**Expected behavior**
Public programs can be visited **and** edited by anyone.

### PR DESCRIPTION
**Pull request by TiBiBa**
**Description**
This PR should fix the bug where a teacher was unable to view any public program.

**Fixes**
This PR fixes #2174 and fixes #2246

**How to test**
Any user should be able to edit any public program. Regardless of being in a class, being a teacher or any other way around. A public program should always be edit-able. In the case of a teacher also a non-public program of one of their students is edit-able.

### CODE REVIEW

**Review by boryanagoncharenko — CHANGES_REQUESTED** [review]
We should address that an admin who is also a teacher cannot see the programs of users who are not their students. @TiBiBa could you have a look at my suggestion in the comment?

File: app.py [hunk 1]
```diff
@@ -864,7 +864,7 @@ def index(level, program_id):
             return utils.error_page(error=404, ui_message=gettext(u'no_such_program'))
 
         # If the current user is a teacher, perform an extra check -> user is their student
-        if is_teacher(user) and result['username'] not in DATABASE.get_teacher_students(user['username']):
+        if is_teacher(user) and not public_program and user['username'] != result['username'] and result['username'] not in DATABASE.get_teacher_students(user['username']):
```
[the remark this answers is not in the document]
Comment (TiBiBa) [reply app.py:867]: I like it! This code is much cleaner and more read-able. Will fix it in the branch, thanks!

**Comment by TiBiBa** [comment]
> @boryanagoncharenko Another weird thing I just noticed is the `@app.route('/hedy/<id>/view/`. This route is called to only show the program, for example when clicking on it in the `/explore` page on when a teacher views programs through a students program overview. This is the point where users click on the edit button and load the program into the editor, the current validation process where are solving in this PR.
> 
> But, on this route we don't perform any user verification if they are "allowed" to view the program. This seems ambiguous. So if I understand correctly the current flow is:
> 
> * Anyone can **see** any program as long as they have the program id
> * But for program editing we perform multiple validation steps to make sure they are allowed
> 
> I think we should align this as well, what do you think?

This should be fixed with the latest commit! We do have some duplicate code however, it might be nice to clean this up a bit and store the "can I access this program" validation in a separate function. But the functionality should be fine.

**Review by boryanagoncharenko — APPROVED** [review]
Agree!

**Pull request merged**